File: server/src/main/kotlin/com/example/reconix/server/service/PdfGeneratorService.kt
package com.example.reconix.server.service

import com.lowagie.text.*
import com.lowagie.text.pdf.PdfPCell
import com.lowagie.text.pdf.PdfPTable
import com.lowagie.text.pdf.PdfWriter
import java.awt.Color
import java.io.File
import java.io.FileOutputStream
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * PDF Generator Service - Creates professional Purchase Order PDFs using OpenPDF
 */
class PdfGeneratorService {

    private val uploadDir = "uploads/po"

    init {
        File(uploadDir).mkdirs()
    }

    /**
     * Generate a Purchase Order PDF document
     * @param poId Purchase Order ID
     * @param vendorName Vendor name
     * @param vendorEmail Vendor email
     * @param items List of line items (itemName, quantity, unitPrice, taxRate)
     * @param totalAmount Total PO amount
     * @return File path of the generated PDF
     */
    fun generatePoPdf(
        poId: String,
        vendorName: String,
        vendorEmail: String,
        items: List<PdfLineItem>,
        totalAmount: Double
    ): String {
        val fileName = "${poId}_${System.currentTimeMillis()}.pdf"
        val filePath = "$uploadDir/$fileName"
        val document = Document(PageSize.A4, 50f, 50f, 50f, 50f)

        try {
            PdfWriter.getInstance(document, FileOutputStream(filePath))
            document.open()

            // ── Company Header ──
            val headerFont = Font(Font.HELVETICA, 24f, Font.BOLD, Color(30, 58, 138))
            val header = Paragraph("RECONIX", headerFont)
            header.alignment = Element.ALIGN_LEFT
            document.add(header)

            val subtitleFont = Font(Font.HELVETICA, 10f, Font.NORMAL, Color(100, 100, 100))
            val subtitle = Paragraph("Automated Invoice Match Validator", subtitleFont)
            subtitle.setSpacingAfter(5f)
            document.add(subtitle)

            // Divider line
            val line = Paragraph("━".repeat(80))
            line.setSpacingAfter(15f)
            document.add(line)

            // ── PO Title ──
            val titleFont = Font(Font.HELVETICA, 18f, Font.BOLD, Color(0, 0, 0))
            val title = Paragraph("PURCHASE ORDER", titleFont)
            title.alignment = Element.ALIGN_CENTER
            title.setSpacingAfter(20f)
            document.add(title)

            // ── PO Details Table ──
            val detailsTable = PdfPTable(2)
            detailsTable.widthPercentage = 100f
            detailsTable.setWidths(floatArrayOf(1f, 2f))

            val labelFont = Font(Font.HELVETICA, 11f, Font.BOLD, Color(55, 65, 81))
            val valueFont = Font(Font.HELVETICA, 11f, Font.NORMAL, Color(17, 24, 39))

            addDetailRow(detailsTable, "PO Number:", poId, labelFont, valueFont)
            addDetailRow(detailsTable, "Date:", LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")), labelFont, valueFont)
            addDetailRow(detailsTable, "Vendor:", vendorName, labelFont, valueFont)
            if (vendorEmail.isNotBlank()) {
                addDetailRow(detailsTable, "Vendor Email:", vendorEmail, labelFont, valueFont)
            }
            addDetailRow(detailsTable, "Status:", "OPEN", labelFont, valueFont)
            detailsTable.setSpacingAfter(25f)
            document.add(detailsTable)

            // ── Line Items Table ──
            val itemsTable = PdfPTable(5)
            itemsTable.widthPercentage = 100f
            itemsTable.setWidths(floatArrayOf(0.5f, 2f, 0.8f, 1f, 1f))

            val headerCellFont = Font(Font.HELVETICA, 10f, Font.BOLD, Color.WHITE)
            val headerBg = Color(30, 58, 138)

            // Table headers
            listOf("#", "Item Description", "Qty", "Unit Price", "Total").forEach { text ->
                val cell = PdfPCell(Phrase(text, headerCellFont))
                cell.backgroundColor = headerBg
                cell.setPadding(8f)
                cell.horizontalAlignment = Element.ALIGN_CENTER
                itemsTable.addCell(cell)
            }

            // Table rows
            val cellFont = Font(Font.HELVETICA, 10f, Font.NORMAL, Color(17, 24, 39))
            items.forEachIndexed { index, item ->
                val rowBg = if (index % 2 == 0) Color(249, 250, 251) else Color.WHITE
                val lineTotal = item.quantity * item.unitPrice

                addItemCell(itemsTable, "${index + 1}", cellFont, rowBg, Element.ALIGN_CENTER)
                addItemCell(itemsTable, item.itemName, cellFont, rowBg, Element.ALIGN_LEFT)
                addItemCell(itemsTable, "${item.quantity}", cellFont, rowBg, Element.ALIGN_CENTER)
                addItemCell(itemsTable, "$${String.format("%.2f", item.unitPrice)}", cellFont, rowBg, Element.ALIGN_RIGHT)
                addItemCell(itemsTable, "$${String.format("%.2f", lineTotal)}", cellFont, rowBg, Element.ALIGN_RIGHT)
            }

            itemsTable.setSpacingAfter(20f)
            document.add(itemsTable)

            // ── Totals Section ──
            val totalsTable = PdfPTable(2)
            totalsTable.widthPercentage = 50f
            totalsTable.horizontalAlignment = Element.ALIGN_RIGHT

            val subtotalAmount = items.sumOf { it.quantity * it.unitPrice }
            val taxAmount = items.sumOf { it.quantity * it.unitPrice * (it.taxRate / 100.0) }

            val totalFont = Font(Font.HELVETICA, 11f, Font.NORMAL)
            val totalBoldFont = Font(Font.HELVETICA, 13f, Font.BOLD, Color(30, 58, 138))

            addTotalRow(totalsTable, "Subtotal:", "$${String.format("%.2f", subtotalAmount)}", totalFont)
            addTotalRow(totalsTable, "Tax:", "$${String.format("%.2f", taxAmount)}", totalFont)

            val grandTotalCell1 = PdfPCell(Phrase("TOTAL:", totalBoldFont))
            grandTotalCell1.border = Rectangle.TOP
            grandTotalCell1.setPadding(8f)
            grandTotalCell1.horizontalAlignment = Element.ALIGN_RIGHT
            totalsTable.addCell(grandTotalCell1)

            val grandTotalCell2 = PdfPCell(Phrase("$${String.format("%.2f", totalAmount)}", totalBoldFont))
            grandTotalCell2.border = Rectangle.TOP
            grandTotalCell2.setPadding(8f)
            grandTotalCell2.horizontalAlignment = Element.ALIGN_RIGHT
            totalsTable.addCell(grandTotalCell2)

            document.add(totalsTable)

            // ── Footer ──
            val footerParagraph = Paragraph()
            footerParagraph.setSpacingBefore(40f)
            val footerFont = Font(Font.HELVETICA, 9f, Font.ITALIC, Color(156, 163, 175))
            footerParagraph.add(Phrase("This is a system-generated document from Reconix Invoice Validator. ", footerFont))
            footerParagraph.add(Phrase("Please reference PO# $poId in all correspondence.", footerFont))
            document.add(footerParagraph)

        } finally {
            document.close()
        }

        println("📄 Generated PO PDF: $filePath")
        return filePath
    }

    private fun addDetailRow(table: PdfPTable, label: String, value: String, labelFont: Font, valueFont: Font) {
        val labelCell = PdfPCell(Phrase(label, labelFont))
        labelCell.border = Rectangle.NO_BORDER
        labelCell.setPadding(4f)
        table.addCell(labelCell)

        val valueCell = PdfPCell(Phrase(value, valueFont))
        valueCell.border = Rectangle.NO_BORDER
        valueCell.setPadding(4f)
        table.addCell(valueCell)
    }

    private fun addItemCell(table: PdfPTable, text: String, font: Font, bgColor: Color, alignment: Int) {
        val cell = PdfPCell(Phrase(text, font))
        cell.backgroundColor = bgColor
        cell.setPadding(6f)
        cell.horizontalAlignment = alignment
        cell.borderColor = Color(229, 231, 235)
        table.addCell(cell)
    }

    private fun addTotalRow(table: PdfPTable, label: String, value: String, font: Font) {
        val labelCell = PdfPCell(Phrase(label, font))
        labelCell.border = Rectangle.NO_BORDER
        labelCell.setPadding(5f)
        labelCell.horizontalAlignment = Element.ALIGN_RIGHT
        table.addCell(labelCell)

        val valueCell = PdfPCell(Phrase(value, font))
        valueCell.border = Rectangle.NO_BORDER
        valueCell.setPadding(5f)
        valueCell.horizontalAlignment = Element.ALIGN_RIGHT
        table.addCell(valueCell)
    }
}

/**
 * Data class for PDF line items
 */
data class PdfLineItem(
    val itemName: String,
    val quantity: Int,
    val unitPrice: Double,
    val taxRate: Double = 0.0
)

File: python_server/services/pdf_generator_service.py
"""
services/pdf_generator_service.py - Generate PO PDFs using ReportLab
Equivalent to server/.../service/PdfGeneratorService.kt
"""
import os
import time
from dataclasses import dataclass, field
from datetime import datetime
from typing import List

UPLOAD_DIR = "uploads/po"
os.makedirs(UPLOAD_DIR, exist_ok=True)


@dataclass
class PdfLineItem:
    item_name: str
    quantity: int
    unit_price: float
    tax_rate: float = 0.0


def generate_po_pdf(
    po_id: str,
    vendor_name: str,
    vendor_email: str,
    items: List[PdfLineItem],
    total_amount: float,
) -> str:
    try:
        from reportlab.lib.pagesizes import A4
        from reportlab.lib import colors
        from reportlab.lib.units import mm
        from reportlab.platypus import (
            SimpleDocTemplate, Table, TableStyle, Paragraph,
            Spacer, HRFlowable,
        )
        from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
        from reportlab.lib.enums import TA_CENTER, TA_RIGHT, TA_LEFT
    except ImportError as e:
        raise RuntimeError(f"reportlab not installed: {e}")

    file_name = f"{po_id}_{int(time.time())}.pdf"
    file_path = os.path.join(UPLOAD_DIR, file_name)

    doc = SimpleDocTemplate(
        file_path,
        pagesize=A4,
        leftMargin=50, rightMargin=50, topMargin=50, bottomMargin=50,
    )

    styles = getSampleStyleSheet()
    dark_blue = colors.HexColor("#1e3a8a")
    light_gray = colors.HexColor("#f9fafb")

    header_style = ParagraphStyle("Header", fontSize=24, textColor=dark_blue, fontName="Helvetica-Bold")
    subtitle_style = ParagraphStyle("Subtitle", fontSize=10, textColor=colors.grey, fontName="Helvetica")
    title_style = ParagraphStyle("Title", fontSize=18, fontName="Helvetica-Bold", alignment=TA_CENTER)
    footer_style = ParagraphStyle("Footer", fontSize=9, fontName="Helvetica-Oblique", textColor=colors.HexColor("#9ca3af"))

    story = []

    # Company header
    story.append(Paragraph("RECONIX", header_style))
    story.append(Paragraph("Automated Invoice Match Validator", subtitle_style))
    story.append(Spacer(1, 8))
    story.append(HRFlowable(width="100%", thickness=1, color=colors.grey))
    story.append(Spacer(1, 12))

    # Document title
    story.append(Paragraph("PURCHASE ORDER", title_style))
    story.append(Spacer(1, 20))

    # Details table
    now_str = datetime.now().strftime("%Y-%m-%d %H:%M")
    detail_rows = [
        ["PO Number:", po_id],
        ["Date:", now_str],
        ["Vendor:", vendor_name],
    ]
    if vendor_email:
        detail_rows.append(["Vendor Email:", vendor_email])
    detail_rows.append(["Status:", "OPEN"])

    detail_table = Table(detail_rows, colWidths=[120, 350])
    detail_table.setStyle(TableStyle([
        ("FONTNAME", (0, 0), (0, -1), "Helvetica-Bold"),
        ("FONTSIZE", (0, 0), (-1, -1), 11),
        ("TEXTCOLOR", (0, 0), (0, -1), colors.HexColor("#374151")),
        ("TEXTCOLOR", (1, 0), (1, -1), colors.HexColor("#111827")),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 5),
        ("TOPPADDING", (0, 0), (-1, -1), 5),
    ]))
    story.append(detail_table)
    story.append(Spacer(1, 20))

    # Items table
    col_headers = ["#", "Item Description", "Qty", "Unit Price", "Total"]
    item_rows = [col_headers]
    for idx, item in enumerate(items):
        line_total = item.quantity * item.unit_price
        item_rows.append([
            str(idx + 1),
            item.item_name,
            str(item.quantity),
            f"${item.unit_price:.2f}",
            f"${line_total:.2f}",
        ])

    items_table = Table(item_rows, colWidths=[30, 230, 60, 80, 80])
    # Alternating row colors
    row_styles = [
        ("BACKGROUND", (0, 0), (-1, 0), dark_blue),
        ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
        ("FONTNAME", (0, 0), (-1, 0), "Helvetica-Bold"),
        ("FONTSIZE", (0, 0), (-1, -1), 10),
        ("ALIGN", (0, 0), (-1, -1), "CENTER"),
        ("ALIGN", (1, 1), (1, -1), "LEFT"),
        ("ALIGN", (3, 1), (-1, -1), "RIGHT"),
        ("GRID", (0, 0), (-1, -1), 0.5, colors.HexColor("#e5e7eb")),
        ("ROWBACKGROUND", (0, 1), (-1, -1), [light_gray, colors.white]),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 6),
        ("TOPPADDING", (0, 0), (-1, -1), 6),
    ]
    items_table.setStyle(TableStyle(row_styles))
    story.append(items_table)
    story.append(Spacer(1, 20))

    # Totals
    subtotal = sum(it.quantity * it.unit_price for it in items)
    tax_amount = sum(it.quantity * it.unit_price * (it.tax_rate / 100.0) for it in items)
    bold_blue = ParagraphStyle("BoldBlue", fontSize=13, fontName="Helvetica-Bold", textColor=dark_blue, alignment=TA_RIGHT)

    totals_data = [
        ["Subtotal:", f"${subtotal:.2f}"],
        ["Tax:", f"${tax_amount:.2f}"],
        [Paragraph("TOTAL:", bold_blue), Paragraph(f"${total_amount:.2f}", bold_blue)],
    ]
    totals_table = Table(totals_data, colWidths=[350, 130], hAlign="RIGHT")
    totals_table.setStyle(TableStyle([
        ("FONTSIZE", (0, 0), (-1, -1), 11),
        ("ALIGN", (0, 0), (-1, -1), "RIGHT"),
        ("LINEABOVE", (0, 2), (-1, 2), 1, colors.grey),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 5),
        ("TOPPADDING", (0, 0), (-1, -1), 5),
    ]))
    story.append(totals_table)
    story.append(Spacer(1, 40))

    # Footer
    story.append(Paragraph(
        f"This is a system-generated document from Reconix Invoice Validator. "
        f"Please reference PO# {po_id} in all correspondence.",
        footer_style,
    ))

    doc.build(story)
    print(f"📄 Generated PO PDF: {file_path}")
    return file_path
